File: scripts/backfill_cn_social_financing_flow.py
#!/usr/bin/env python3
"""Backfill CN Social Financing Flow into Supabase macro_cn.

- Source: AkShare macro_china_shrzgm() (monthly flow, unit typically 100 million RMB)
- Target: Supabase table macro_cn column social_financing_flow (numeric)
- Strategy: write to month-end date rows; upsert only (date + social_financing_flow + source + updated_at)
- IMPORTANT: does not overwrite other macro_cn fields.

Usage:
  SUPABASE_URL=... SUPABASE_SERVICE_ROLE_KEY=... python3 scripts/backfill_cn_social_financing_flow.py

"""

import os
import sys
import datetime as dt
import pandas as pd

try:
    import akshare as ak
except Exception as e:
    print("ERROR: akshare not available:", e)
    sys.exit(1)

try:
    from supabase import create_client
except Exception:
    print("ERROR: python supabase client not installed. Run: pip install supabase")
    sys.exit(1)

SUPABASE_URL = os.getenv("SUPABASE_URL") or os.getenv("NEXT_PUBLIC_SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")

if not SUPABASE_URL or not SUPABASE_KEY:
    print("ERROR: Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY")
    sys.exit(2)


def month_end_date(yyyymm: str) -> str:
    y = int(yyyymm[:4])
    m = int(yyyymm[4:6])
    # month end: first day next month - 1 day
    if m == 12:
        ny, nm = y + 1, 1
    else:
        ny, nm = y, m + 1
    d = dt.date(ny, nm, 1) - dt.timedelta(days=1)
    return d.isoformat()


def main():
    print("[CN][SHRZGM] fetching ak.macro_china_shrzgm() …")
    df = ak.macro_china_shrzgm()
    if df is None or df.empty:
        raise RuntimeError("AkShare returned empty dataframe")

    if "月份" not in df.columns or "社会融资规模增量" not in df.columns:
        raise RuntimeError(f"Unexpected columns: {list(df.columns)}")

    out = df[["月份", "社会融资规模增量"]].copy()
    out["month"] = out["月份"].astype(str).str.slice(0, 6)
    out["value"] = pd.to_numeric(out["社会融资规模增量"], errors="coerce")
    out = out.dropna(subset=["value", "month"])

    now = dt.datetime.utcnow().replace(microsecond=0).isoformat() + "Z"

    rows = []
    for _, r in out.iterrows():
        yyyymm = str(r["month"])
        d = month_end_date(yyyymm)
        rows.append(
            {
                "date": f"{d}T00:00:00+00:00",
                "social_financing_flow": float(r["value"]),
                "source": "akshare_shrzgm",
                "updated_at": now,
            }
        )

    print(f"[CN][SHRZGM] months={len(rows)} range={rows[0]['date'][:10]}..{rows[-1]['date'][:10]}")

    sb = create_client(SUPABASE_URL, SUPABASE_KEY)

    # upsert in batches
    batch = 200
    for i in range(0, len(rows), batch):
        part = rows[i : i + batch]
        resp = sb.table("macro_cn").upsert(part, on_conflict="date").execute()
        # supabase-py returns data+count; errors raise
        _ = resp

    print("Done.")


if __name__ == "__main__":
    main()

File: scripts/backfill_cn_social_financing_flow.mjs
import akshare from "akshare";
import { createClient } from "@supabase/supabase-js";

const SUPABASE_URL = process.env.SUPABASE_URL || process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

if (!SUPABASE_URL || !SUPABASE_KEY) {
  console.error("Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY/NEXT_PUBLIC_SUPABASE_ANON_KEY");
  process.exit(1);
}

function monthEnd(yyyymm) {
  const y = Number(yyyymm.slice(0, 4));
  const m = Number(yyyymm.slice(4, 6));
  const d = new Date(Date.UTC(y, m, 0)).getUTCDate();
  return `${yyyymm.slice(0, 4)}-${yyyymm.slice(4, 6)}-${String(d).padStart(2, "0")}`;
}

function toNum(x) {
  const n = Number(String(x).trim());
  return Number.isFinite(n) ? n : null;
}

async function main() {
  console.log("[CN][SHRZGM] fetching AkShare macro_china_shrzgm() …");
  const df = await akshare.macro_china_shrzgm();
  // akshare node returns array of objects
  if (!Array.isArray(df) || df.length === 0) {
    throw new Error("AkShare returned empty dataset");
  }

  const rows = [];
  for (const r of df) {
    const ym = String(r["月份"] ?? "").slice(0, 6);
    const v = toNum(r["社会融资规模增量"]);
    if (!ym || v === null) continue;
    const d = monthEnd(ym);
    rows.push({
      date: `${d}T00:00:00+00:00`,
      social_financing_flow: v,
      source: "akshare_shrzgm",
      updated_at: new Date().toISOString(),
    });
  }

  rows.sort((a, b) => (a.date < b.date ? -1 : 1));
  console.log(`[CN][SHRZGM] months=${rows.length} range=${rows[0]?.date?.slice(0, 10)}..${rows[rows.length - 1]?.date?.slice(0, 10)}`);

  const sb = createClient(SUPABASE_URL, SUPABASE_KEY);

  const BATCH = 200;
  for (let i = 0; i < rows.length; i += BATCH) {
    const part = rows.slice(i, i + BATCH);
    const { error } = await sb.from("macro_cn").upsert(part, { onConflict: "date" });
    if (error) throw new Error(error.message);
  }

  console.log("Done.");
}

main().catch((e) => {
  console.error("FAILED:", e);
  process.exit(2);
});
